pkg/billing/azureblobstoragebilling.go

```diff
@@ -0,0 +1,143 @@
+	Container string `json:"container"`
+	// Files will be created with name starting with PrefixInContainer
+	PrefixInContainer string `json:"prefixInContainer"`
+	// Example Endpoint: "https://MYSTORAGEACCOUNT.blob.core.windows.net/"
+	Endpoint string `json:"endpoint"`
+
+	//
+	// Unexported attributes follow this comment.
+	//
+
+	// Use generateKey for tests.
+	// Otherwise, keep empty.
+	generateKey func() string
+}
+
+type AzureError struct {
+	Err error
+}
+
+func (e AzureError) Error() string {
+	return fmt.Sprintf("Azure Blob error: %s", e.Err.Error())
+}
+
+func (e AzureError) Unwrap() error {
+	return e.Err
+}
+
+type AzureClient struct {
+	cfg AzureBlobStorageClientConfig
+	c   *azblob.Client
+}
+
+func (c AzureClient) LogFields() zap.Field {
+	return zap.Inline(zapcore.ObjectMarshalerFunc(func(enc zapcore.ObjectEncoder) error {
+		enc.AddString("container", c.cfg.Container)
+		enc.AddString("prefixInContainer", c.cfg.PrefixInContainer)
+		enc.AddString("endpoint", c.cfg.Endpoint)
+		return nil
+	}))
+}
+
+func (c AzureClient) generateKey() string {
+	if c.cfg.generateKey != nil {
+		return c.cfg.generateKey()
+	}
+	return keyTemplate(c.cfg.PrefixInContainer)
+}
+
+func (c AzureClient) send(ctx context.Context, payload []byte, _ TraceID) error {
+	payload, err := bytesForBilling(payload)
+	if err != nil {
+		return err
+	}
+	_, err = c.c.UploadBuffer(ctx, c.cfg.Container, c.generateKey(), payload,
+		&azblob.UploadBufferOptions{}, //nolint:exhaustruct // It's part of Azure SDK
+	)
+	return handleAzureError(err)
+}
+
+func NewAzureBlobStorageClient(cfg AzureBlobStorageClientConfig) (*AzureClient, error) {
+	var client *azblob.Client
+
+	//nolint:exhaustruct // It's part of Azure SDK
+	clientOptions := &azblob.ClientOptions{
+		ClientOptions: azcore.ClientOptions{
+			Telemetry: policy.TelemetryOptions{ApplicationID: "neon-autoscaler"},
+		},
+	}
+	switch cfg.AuthType {
+	case azureAuthTypeTests:
+		// Using well known credentials,
+		// see https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azurite
+		shKey, err := azblob.NewSharedKeyCredential(
+			"devstoreaccount1",
+			"Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==")
+		if err != nil {
+			return nil, err
+		}
+
+		client, err = azblob.NewClientWithSharedKeyCredential(cfg.Endpoint, shKey, clientOptions)
+		if err != nil {
+			return nil, &AzureError{err}
+		}
+	case AzureAuthTypeDefault:
+		credential, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, err
+		}
+		client, err = azblob.NewClient(cfg.Endpoint, credential, clientOptions)
+		if err != nil {
+			return nil, &AzureError{err}
+		}
+	default:
+		return nil, fmt.Errorf("unsupported auth type: %q", cfg.AuthType)
+	}
+
+	return &AzureClient{
+		cfg: cfg,
+		c:   client,
+	}, nil
+}
+
+func handleAzureError(err error) error {
+	if err == nil {
+		return nil
+	}
+	return AzureError{err}
```

Comment:
The reason those error wrappers exist is to reduce the cardinality for the errors graph.

https://github.com/neondatabase/autoscaling/blob/8195df03ebe55f1dc185965a7f0649b9edd6e0bb/pkg/agent/billing/send.go#L154

I think the same should be done for azure.